lib/CL/pocl_llvm_dynlib.cc

```diff
@@ -65,34 +66,87 @@ void *pocl_dynlib_symbol_address(void *, const char *SymbolName) {
 
 const char *pocl_dynlib_pathname(void *Address) {
 #ifdef _WIN32
+  // Look up the module handle from the address
   HMODULE Hm;
   if (GetModuleHandleExW(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS |
                              GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
-                         (LPCWSTR)&Address, &Hm) == 0) {
+                         (LPCWSTR)Address, &Hm) == 0) {
     int Ret = GetLastError();
-    POCL_MSG_WARN("GetModuleHandleEx failed, error = %d ; trying fallback\n", Ret);
+    POCL_MSG_WARN("GetModuleHandleEx failed, error = %d; trying fallback\n",
+                  Ret);
 
-    // undocumented hack from https://stackoverflow.com/a/2396380
+    // Undocumented hack from https://stackoverflow.com/a/2396380
     MEMORY_BASIC_INFORMATION mbi;
-    size_t Len = VirtualQuery(Address, &mbi, sizeof(mbi));
-    if (Len != sizeof(mbi)) {
-      POCL_MSG_ERR("VirtualQuery failed", Ret);
+    if (VirtualQuery(Address, &mbi, sizeof(mbi)) == 0) {
+      Ret = GetLastError();
+      POCL_MSG_ERR("VirtualQuery fallback failed, error = %d\n", Ret);
       return nullptr;
     }
     Hm = (HMODULE)mbi.AllocationBase;
   }
 
-  WCHAR wpath[MAX_PATH]{};
+  // Get the path of the module
+  WCHAR wpath[MAX_PATH];
   if (GetModuleFileNameW(Hm, wpath, ARRAYSIZE(wpath)) == 0) {
     int Ret = GetLastError();
     POCL_MSG_ERR("GetModuleFileName failed, error = %d\n", Ret);
     return nullptr;
   }
 
-  static char path[MAX_PATH];
-  WideCharToMultiByte(CP_UTF8, 0, wpath, -1, path, ARRAYSIZE(path), NULL, NULL);
-  POCL_MSG_PRINT_INFO("pocl_dynlib_symbol_adress: using DLL path: %s \n", path);
-  return path;
+  // Open a the file to get a handle
+  HANDLE hFile =
+      CreateFileW(wpath, GENERIC_READ,
+                  FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
+                  OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+  if (hFile == INVALID_HANDLE_VALUE) {
+    int Ret = GetLastError();
+    POCL_MSG_ERR("CreateFile failed, error = %d\n", Ret);
+    return nullptr;
+  }
+
+  // Get the final path, resolving any symlinks
+  std::vector<WCHAR> wfinal_buf(MAX_PATH);
+  DWORD path_len = GetFinalPathNameByHandleW(
+      hFile, wfinal_buf.data(), wfinal_buf.size(), VOLUME_NAME_DOS);
```

Comment:
Could this call fail (aka. call return zero) because the symlink points to s non-existent file?

Reply:
Ah, never mind - this case is covered.

Reply:
Given that we got the path from looking it up from the current module, that seems unlikely?

Reply:
That said, there's other ways this could fail (e.g. by passing invalid arguments), so it's probably best to add an error check. I'll update the PR.